File: src/main/java/mods/eln/sim/mna/component/ResistorSwitch.java
package mods.eln.sim.mna.component;

import mods.eln.misc.INBTTReady;
import mods.eln.sim.mna.misc.MnaConst;
import mods.eln.sim.mna.state.State;
import net.minecraft.nbt.NBTTagCompound;

public class ResistorSwitch extends Resistor implements INBTTReady {

    String name;

    private boolean state = false;

    protected double baseResistance = 1;

    protected double offResistance = MnaConst.highImpedance;

    public ResistorSwitch(String name, State aPin, State bPin) {
        super(aPin, bPin);
        this.name = name;
    }

    public void setState(boolean state) {
        this.state = state;
        super.setResistance(state ? baseResistance : offResistance);
    }

    public void setOffResistance(double resistance) {
        offResistance = resistance;
        super.setResistance(state ? baseResistance : offResistance);
    }

    @Override
    public void highImpedance() {
        super.setResistance(offResistance);
    }

    @Override
    public Resistor setResistance(double resistance) {
        baseResistance = resistance;
        return super.setResistance(state ? resistance : offResistance);
    }

    public boolean getState() {
        return state;
    }

    @Override
    public void readFromNBT(NBTTagCompound nbt, String str) {
        str += name;
        double resistance = nbt.getDouble(str + "R");
        if (!Double.isFinite(resistance) || resistance == 0) {
            baseResistance = offResistance;
        } else {
            baseResistance = resistance;
        }
        state = nbt.getBoolean(str + "State");
        super.setResistance(state ? baseResistance : offResistance);
    }

    @Override
    public void writeToNBT(NBTTagCompound nbt, String str) {
        str += name;
        nbt.setDouble(str + "R", baseResistance);
        nbt.setBoolean(str + "State", getState());
    }
}

File: src/main/java/mods/eln/sim/mna/component/ResistorSwitch.kt
package mods.eln.sim.mna.component

import mods.eln.misc.INBTTReady
import mods.eln.sim.mna.misc.MnaConst
import mods.eln.sim.mna.state.State
import net.minecraft.nbt.NBTTagCompound

class ResistorSwitch : Resistor, INBTTReady {

    constructor(name: String, aPin: State?, bPin: State?) {
        this.name = name
        connectTo(aPin, bPin)
    }

    internal var ultraImpedance = false

    internal var state = false

    override var r = MnaConst.highImpedance
        set(r) {
            baseR = r
            field = if (state) r else if (ultraImpedance) MnaConst.ultraImpedance else MnaConst.highImpedance
        }

    protected var baseR = 1.0

    fun setState(state: Boolean) {
        this.state = state
        r = baseR
    }

    fun getState(): Boolean {
        return state
    }

    override fun readFromNBT(nbt: NBTTagCompound, str: String) {
        var strl = str
        strl += name
        r = nbt.getDouble(strl + "R")
        if (java.lang.Double.isNaN(baseR) || baseR == 0.0) {
            if (ultraImpedance)
                ultraImpedance()
            else
                highImpedance()
        }
        setState(nbt.getBoolean(strl + "State"))
    }

    override fun writeToNBT(nbt: NBTTagCompound, str: String) {
        var strl = str
        strl += name
        nbt.setDouble(strl + "R", baseR)
        nbt.setBoolean(strl + "State", getState())
    }

    fun mustUseUltraImpedance() {
        ultraImpedance = true
    }
}
